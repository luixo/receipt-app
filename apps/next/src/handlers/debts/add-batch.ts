import { TRPCError } from "@trpc/server";
import type { DatabaseError } from "pg";
import { z } from "zod";

import { nonNullishGuard } from "app/utils/utils";
import {
	MAX_BATCH_DEBTS,
	MIN_BATCH_DEBTS,
	debtAmountSchema,
	debtNoteSchema,
} from "app/utils/validation";
import type { DebtsId, UsersId } from "next-app/db/models";
import { authProcedure } from "next-app/handlers/trpc";
import {
	currencyCodeSchema,
	receiptIdSchema,
	userIdSchema,
} from "next-app/handlers/validation";

import { withOwnerReceiptUserConstraint } from "./utils";

const extractError = (e: unknown) => {
	/* c8 ignore start */
	if (
		typeof e !== "object" ||
		!e ||
		!("detail" in e) ||
		typeof e.detail !== "string"
	) {
		throw new TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Error is not of type 'DatabaseError': ${String(e)}`,
		});
	}
	/* c8 ignore stop */
	const typedMessage = e as DatabaseError;
	const detailMatch = typedMessage.detail!.match(
		/\("ownerAccountId", "receiptId", "userId"\)=\([a-z0-9-]+, ([a-z0-9-]+), ([a-z0-9-]+)\) already exists/,
	);
	/* c8 ignore start */
	if (!detailMatch) {
		throw new TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Detail didn't match expected constraint message: ${typedMessage.detail}`,
		});
	}
	/* c8 ignore stop */
	const [, receiptId, userId] = detailMatch;
	return { receiptId: receiptId!, userId: userId! };
};

export const procedure = authProcedure
	.input(
		z
			.strictObject({
				note: debtNoteSchema,
				currencyCode: currencyCodeSchema,
				userId: userIdSchema,
				amount: debtAmountSchema,
				timestamp: z.date().optional(),
				receiptId: receiptIdSchema.optional(),
			})
			.array()
			.max(
				MAX_BATCH_DEBTS,
				`Maximum amount of batched debts is ${MAX_BATCH_DEBTS}`,
			)
			.min(
				MIN_BATCH_DEBTS,
				`Minimal amount of batched debts is ${MIN_BATCH_DEBTS}`,
			),
	)
	.mutation(async ({ input: debts, ctx }) => {
		const ids: DebtsId[] = debts.map(() => ctx.getUuid());
		const { database } = ctx;
		const userIds = [...new Set(debts.map(({ userId }) => userId))];
		const users = await database
			.selectFrom("users")
			.leftJoin("accountSettings", (qb) =>
				qb.onRef("users.connectedAccountId", "=", "accountSettings.accountId"),
			)
			.select([
				"users.id",
				"users.ownerAccountId as selfAccountId",
				"accountSettings.accountId as foreignAccountId",
				"accountSettings.autoAcceptDebts",
			])
			.where("users.id", "in", userIds)
			.execute();
		if (users.length !== userIds.length) {
			const foundUserIds = users.map((user) => user.id);
			const missingUserIds = userIds.filter(
				(requestedUserId) => !foundUserIds.includes(requestedUserId),
			);
			throw new TRPCError({
				code: "NOT_FOUND",
				message: `${
					missingUserIds.length === 1 ? "User" : "Users"
				} ${missingUserIds.map((userId) => `"${userId}"`).join(", ")} ${
					missingUserIds.length === 1 ? "does" : "do"
				} not exist.`,
			});
		}
		const foreignUsers = users.filter(
			({ selfAccountId }) => selfAccountId !== ctx.auth.accountId,
		);
		if (foreignUsers.length !== 0) {
			const foreignUserIds = foreignUsers.map((foreignUser) => foreignUser.id);
			throw new TRPCError({
				code: "FORBIDDEN",
				message: `${
					foreignUserIds.length === 1 ? "User" : "Users"
				} ${foreignUserIds
					.sort()
					.map((userId) => `"${userId}"`)
					.join(", ")} ${
					foreignUserIds.length === 1 ? "is" : "are"
				} not owned by "${ctx.auth.email}".`,
			});
		}
		const hasSelfUser = users.some((user) => user.id === user.selfAccountId);
		if (hasSelfUser) {
			throw new TRPCError({
				code: "FORBIDDEN",
				message: `Cannot add debts for yourself.`,
			});
		}
		const autoAcceptingUsers = users.filter((user) => user.autoAcceptDebts);
		const lockedTimestamp = new Date();
		let reverseAcceptedUserIds: UsersId[] = [];
		const commonParts = debts.map((debt, index) => ({
			id: ids[index]!,
			note: debt.note,
			currencyCode: debt.currencyCode,
			created: new Date(),
			timestamp: debt.timestamp || new Date(),
			lockedTimestamp,
			receiptId: debt.receiptId,
		}));
		if (autoAcceptingUsers.length !== 0) {
			const autoAcceptingUsersAccountIds = autoAcceptingUsers
				.map((user) => user.foreignAccountId)
				.filter(nonNullishGuard);
			/* c8 ignore start */
			if (autoAcceptingUsersAccountIds.length !== autoAcceptingUsers.length) {
				throw new TRPCError({
					code: "INTERNAL_SERVER_ERROR",
					message: `Unexpected having "autoAcceptDebts" but not having "accountId"`,
				});
			}
			/* c8 ignore stop */
			// TODO: incorporate reverse user into VALUES clause
			const reverseAutoAcceptingUsers = await database
				.selectFrom("users")
				.where("users.ownerAccountId", "in", autoAcceptingUsersAccountIds)
				.where("users.connectedAccountId", "=", ctx.auth.accountId)
				.select(["users.ownerAccountId", "users.id"])
				.execute();
			/* c8 ignore start */
			if (
				reverseAutoAcceptingUsers.length !== autoAcceptingUsersAccountIds.length
			) {
				throw new TRPCError({
					code: "INTERNAL_SERVER_ERROR",
					message: `Unexpected having "autoAcceptDebts" but not having reverse user "id"`,
				});
			}
			/* c8 ignore stop */
			await withOwnerReceiptUserConstraint(
				() =>
					database
						.insertInto("debts")
						.values(
							commonParts
								.map((commonPart, index) => {
									const matchedDebt = debts[index]!;
									const matchedUser = users.find(
										(user) => matchedDebt.userId === user.id,
									)!;
									const matchedReverseUser = reverseAutoAcceptingUsers.find(
										(reverseUser) =>
											reverseUser.ownerAccountId ===
											matchedUser.foreignAccountId,
									);
									if (!matchedReverseUser) {
										return null;
									}
									return {
										...commonPart,
										ownerAccountId: matchedReverseUser.ownerAccountId,
										userId: matchedReverseUser.id,
										amount: (-debts[index]!.amount).toString(),
									};
								})
								.filter(nonNullishGuard),
						)
						.executeTakeFirst(),
				extractError,
			);
			reverseAcceptedUserIds = autoAcceptingUsers.map((user) => user.id);
		}
		await withOwnerReceiptUserConstraint(
			() =>
				database
					.insertInto("debts")
					.values(
						commonParts.map((commonPart, index) => ({
							...commonPart,
							ownerAccountId: ctx.auth.accountId,
							userId: debts[index]!.userId,
							amount: debts[index]!.amount.toString(),
						})),
					)
					.execute(),
			extractError,
		);
		return { ids, lockedTimestamp, reverseAcceptedUserIds };
	});
