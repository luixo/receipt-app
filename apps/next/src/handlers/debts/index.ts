import { t } from "next-app/handlers/trpc";

import { procedure as add } from "./add";
import { procedure as addBatch } from "./add-batch";
import { procedure as get } from "./get";
import { procedure as getByReceiptId } from "./get-by-receipt-id";
import { procedure as getByUsers } from "./get-by-users";
import { procedure as getReceipt } from "./get-receipt";
import { procedure as getUser } from "./get-user";
import { procedure as remove } from "./remove";
import { procedure as update } from "./update";

export const router = t.router({
	getByUsers,
	getUser,
	getReceipt,
	get,
	update,
	remove,
	add,
	addBatch,
	getByReceiptId,
});
