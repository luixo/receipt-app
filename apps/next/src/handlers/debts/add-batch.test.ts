import { faker } from "@faker-js/faker";
import { describe, expect } from "vitest";

import { createAuthContext } from "@tests/backend/utils/context";
import {
	insertAccount,
	insertAccountSettings,
	insertAccountWithSession,
	insertConnectedUsers,
	insertDebt,
	insertReceipt,
	insertReceiptParticipant,
	insertUser,
} from "@tests/backend/utils/data";
import {
	expectDatabaseDiffSnapshot,
	expectTRPCError,
	expectUnauthorizedError,
} from "@tests/backend/utils/expect";
import { test } from "@tests/backend/utils/test";
import { MAX_BATCH_DEBTS, MIN_BATCH_DEBTS } from "app/utils/validation";
import { t } from "next-app/handlers/trpc";
import { UUID_REGEX } from "next-app/handlers/validation";

import { procedure } from "./add-batch";
import {
	getValidDebt,
	verifyAmount,
	verifyCurrencyCode,
	verifyNote,
	verifyReceiptId,
	verifyTimestamp,
	verifyUserId,
} from "./utils.test";

const router = t.router({ procedure });

describe("debts.addBatch", () => {
	describe("input verification", () => {
		expectUnauthorizedError((context) =>
			router.createCaller(context).procedure([getValidDebt()]),
		);

		describe("input array", () => {
			test("minimal length", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx);
				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() => caller.procedure([]),
					"BAD_REQUEST",
					`Zod error\n\nAt "<root>": Minimal amount of batched debts is ${MIN_BATCH_DEBTS}`,
				);
			});

			test("maximum length", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx);
				const caller = router.createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure(
							Array.from({ length: MAX_BATCH_DEBTS + 1 }, () => getValidDebt()),
						),
					"BAD_REQUEST",
					`Zod error\n\nAt "<root>": Maximum amount of batched debts is ${MAX_BATCH_DEBTS}`,
				);
			});
		});

		verifyAmount(
			(context, amount) =>
				router.createCaller(context).procedure([{ ...getValidDebt(), amount }]),
			"[0].",
		);

		verifyNote(
			(context, note) =>
				router.createCaller(context).procedure([{ ...getValidDebt(), note }]),
			"[0].",
		);

		verifyCurrencyCode(
			(context, currencyCode) =>
				router
					.createCaller(context)
					.procedure([{ ...getValidDebt(), currencyCode }]),
			"[0].",
		);

		verifyTimestamp(
			(context, timestamp) =>
				router
					.createCaller(context)
					.procedure([{ ...getValidDebt(), timestamp }]),
			"[0].",
		);

		verifyReceiptId(
			(context, receiptId) =>
				router
					.createCaller(context)
					.procedure([{ ...getValidDebt(), receiptId }]),
			"[0].",
		);

		verifyUserId(
			(context, userId) =>
				router.createCaller(context).procedure([getValidDebt(userId)]),
			"[0].",
		);

		test("users do not exist", async ({ ctx }) => {
			const { sessionId } = await insertAccountWithSession(ctx);
			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			const fakeUserId = faker.string.uuid();
			const anotherFakeUserId = faker.string.uuid();
			await expectTRPCError(
				() => caller.procedure([getValidDebt(fakeUserId)]),
				"NOT_FOUND",
				`User "${fakeUserId}" does not exist.`,
			);
			await expectTRPCError(
				() =>
					caller.procedure([
						getValidDebt(fakeUserId),
						getValidDebt(anotherFakeUserId),
					]),
				"NOT_FOUND",
				`Users "${fakeUserId}", "${anotherFakeUserId}" do not exist.`,
			);
		});

		test("users are not owned by an account", async ({ ctx }) => {
			const { sessionId, accountId, account } = await insertAccountWithSession(
				ctx,
			);
			await insertUser(ctx, accountId);

			const { id: foreignAccountId } = await insertAccount(ctx);
			const { id: foreignUserId } = await insertUser(ctx, foreignAccountId);
			const { id: anotherForeignUserId } = await insertUser(
				ctx,
				foreignAccountId,
			);

			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() => caller.procedure([getValidDebt(foreignUserId)]),
				"FORBIDDEN",
				`User "${foreignUserId}" is not owned by "${account.email}".`,
			);
			await expectTRPCError(
				() =>
					caller.procedure([
						getValidDebt(foreignUserId),
						getValidDebt(anotherForeignUserId),
					]),
				"FORBIDDEN",
				`Users "${foreignUserId}", "${anotherForeignUserId}" are not owned by "${account.email}".`,
			);
		});

		test("user is ourselves", async ({ ctx }) => {
			const { sessionId, userId } = await insertAccountWithSession(ctx);

			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() => caller.procedure([getValidDebt(userId)]),
				"FORBIDDEN",
				`Cannot add debts for yourself.`,
			);
		});

		test("there is a receipt id debt with that user", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const { id: receiptId } = await insertReceipt(ctx, accountId);
			const { id: userId } = await insertUser(ctx, accountId);
			await insertReceiptParticipant(ctx, receiptId, userId);
			await insertDebt(ctx, accountId, userId, { receiptId });

			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure([
						{
							...getValidDebt(userId),
							receiptId,
						},
					]),
				"FORBIDDEN",
				`There is already a debt for user "${userId}" in receipt "${receiptId}".`,
			);
		});
	});

	describe("functionality", () => {
		test("debts are added - autogenerated timestamp", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const { id: userId } = await insertUser(ctx, accountId);
			const { id: anotherUserId } = await insertUser(ctx, accountId);

			// Verify unrelated data doesn't affect the result
			await insertUser(ctx, accountId);
			await insertAccountSettings(ctx, accountId, { autoAcceptDebts: true });
			const { id: foreignAccountId } = await insertAccount(ctx);
			await insertUser(ctx, foreignAccountId);

			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			const result = await expectDatabaseDiffSnapshot(ctx, () =>
				caller.procedure([getValidDebt(userId), getValidDebt(anotherUserId)]),
			);
			expect(result.ids).toHaveLength(2);
			result.ids.every((id) => expect(id).toMatch(UUID_REGEX));
			expect(result).toStrictEqual<typeof result>({
				ids: result.ids,
				lockedTimestamp: new Date(),
				reverseAcceptedUserIds: [],
			});
		});

		test("debts are added - user-provided timestamp", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const { id: userId } = await insertUser(ctx, accountId);
			const { id: anotherUserId } = await insertUser(ctx, accountId);

			// Verify unrelated data doesn't affect the result
			await insertUser(ctx, accountId);
			await insertAccountSettings(ctx, accountId, { autoAcceptDebts: true });
			const { id: foreignAccountId } = await insertAccount(ctx);
			await insertUser(ctx, foreignAccountId);

			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			const result = await expectDatabaseDiffSnapshot(ctx, () =>
				caller.procedure([
					{ ...getValidDebt(userId), timestamp: new Date("2021-01-01") },
					{ ...getValidDebt(anotherUserId), timestamp: new Date("2021-01-01") },
				]),
			);
			expect(result.ids).toHaveLength(2);
			result.ids.every((id) => expect(id).toMatch(UUID_REGEX));
			expect(result).toStrictEqual<typeof result>({
				ids: result.ids,
				lockedTimestamp: new Date(),
				reverseAcceptedUserIds: [],
			});
		});

		test("debts are added - some are auto-accepted by the counterparty", async ({
			ctx,
		}) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const { id: acceptingForeignAccountId } = await insertAccount(ctx, {
				settings: { autoAcceptDebts: true },
			});
			const { id: anotherAcceptingForeignAccountId } = await insertAccount(
				ctx,
				{ settings: { autoAcceptDebts: true } },
			);
			const { id: otherForeignAccountId } = await insertAccount(ctx);

			const [{ id: acceptingUserId }] = await insertConnectedUsers(ctx, [
				accountId,
				acceptingForeignAccountId,
			]);
			const [{ id: anotherAcceptingUserId }] = await insertConnectedUsers(ctx, [
				accountId,
				anotherAcceptingForeignAccountId,
			]);
			const [{ id: otherUserId }] = await insertConnectedUsers(ctx, [
				accountId,
				otherForeignAccountId,
			]);
			const { id: receiptId } = await insertReceipt(ctx, accountId);

			// Verify unrelated data doesn't affect the result
			await insertUser(ctx, accountId);
			await insertAccountSettings(ctx, accountId, { autoAcceptDebts: true });
			await insertUser(ctx, acceptingForeignAccountId);
			await insertUser(ctx, anotherAcceptingForeignAccountId);

			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			const result = await expectDatabaseDiffSnapshot(ctx, () =>
				caller.procedure([
					{ ...getValidDebt(acceptingUserId), receiptId },
					getValidDebt(acceptingUserId),
					getValidDebt(anotherAcceptingUserId),
					getValidDebt(otherUserId),
				]),
			);
			expect(result.ids).toHaveLength(4);
			result.ids.every((id) => expect(id).toMatch(UUID_REGEX));
			expect(result).toStrictEqual<typeof result>({
				ids: result.ids,
				lockedTimestamp: new Date(),
				reverseAcceptedUserIds: [acceptingUserId, anotherAcceptingUserId],
			});
		});

		test("debts are added - for the same auto-accepting account", async ({
			ctx,
		}) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const { id: acceptingForeignAccountId } = await insertAccount(ctx, {
				settings: { autoAcceptDebts: true },
			});

			const [{ id: acceptingUserId }] = await insertConnectedUsers(ctx, [
				accountId,
				acceptingForeignAccountId,
			]);

			// Verify unrelated data doesn't affect the result
			await insertUser(ctx, accountId);
			await insertAccountSettings(ctx, accountId, { autoAcceptDebts: true });
			await insertUser(ctx, acceptingForeignAccountId);

			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			const result = await expectDatabaseDiffSnapshot(ctx, () =>
				caller.procedure([
					getValidDebt(acceptingUserId),
					getValidDebt(acceptingUserId),
					getValidDebt(acceptingUserId),
				]),
			);
			expect(result.ids).toHaveLength(3);
			result.ids.every((id) => expect(id).toMatch(UUID_REGEX));
			expect(result).toStrictEqual<typeof result>({
				ids: result.ids,
				lockedTimestamp: new Date(),
				reverseAcceptedUserIds: [acceptingUserId],
			});
		});

		test("debts are added - some auto-accepting accounts already had debts picked by the same receipt id", async ({
			ctx,
		}) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const { id: acceptingForeignAccountId } = await insertAccount(ctx, {
				settings: { autoAcceptDebts: true },
			});

			const [{ id: acceptingUserId }, { id: selfUserId }] =
				await insertConnectedUsers(ctx, [accountId, acceptingForeignAccountId]);

			const { id: receiptId } = await insertReceipt(
				ctx,
				acceptingForeignAccountId,
			);
			const { id: counterpartyId } = await insertDebt(
				ctx,
				acceptingForeignAccountId,
				selfUserId,
				{ receiptId },
			);

			// Verify unrelated data doesn't affect the result
			await insertUser(ctx, accountId);
			await insertAccountSettings(ctx, accountId, { autoAcceptDebts: true });
			await insertUser(ctx, acceptingForeignAccountId);

			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			const result = await expectDatabaseDiffSnapshot(ctx, () =>
				caller.procedure([
					{ ...getValidDebt(acceptingUserId), receiptId },
					getValidDebt(acceptingUserId),
					getValidDebt(acceptingUserId),
				]),
			);
			expect(result.ids).toHaveLength(3);
			result.ids.every((id) => expect(id).toMatch(UUID_REGEX));
			expect(result).toStrictEqual<typeof result>({
				ids: result.ids,
				lockedTimestamp: new Date(),
				reverseAcceptedUserIds: [acceptingUserId],
			});
			expect(result.ids[0]).toEqual(counterpartyId);
		});
	});
});
