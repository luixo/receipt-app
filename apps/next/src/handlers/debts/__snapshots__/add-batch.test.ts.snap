// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`debts.addBatch > functionality > debts are added - autogenerated timestamp 1`] = `
{
  "added": {
    "debts": {
      "id:5b7ab2ba-be55-4771-8cd0-24e8fe338bd6|ownerAccountId:a4211316-9cb2-4005-b661-283ef7f1cf45": {
        "amount": "855.3800",
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "EGP",
        "id": "5b7ab2ba-be55-4771-8cd0-24e8fe338bd6",
        "lockedTimestamp": "2020-01-01T00:00:00.000Z",
        "note": "occaecati ut minus",
        "ownerAccountId": "a4211316-9cb2-4005-b661-283ef7f1cf45",
        "receiptId": null,
        "timestamp": "2020-01-01T00:00:00.000Z",
        "userId": "9dfe9df8-4ed0-4d14-904b-75ddf3bab1a8",
      },
      "id:845cb5ab-842b-4fa9-acbf-1a759ff393a4|ownerAccountId:a4211316-9cb2-4005-b661-283ef7f1cf45": {
        "amount": "-907.8400",
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "NOK",
        "id": "845cb5ab-842b-4fa9-acbf-1a759ff393a4",
        "lockedTimestamp": "2020-01-01T00:00:00.000Z",
        "note": "debitis culpa ea",
        "ownerAccountId": "a4211316-9cb2-4005-b661-283ef7f1cf45",
        "receiptId": null,
        "timestamp": "2020-01-01T00:00:00.000Z",
        "userId": "354b1a6f-df84-4caf-9e40-14e29dd617ee",
      },
    },
  },
}
`;

exports[`debts.addBatch > functionality > debts are added - for the same auto-accepting account 1`] = `
{
  "added": {
    "debts": {
      "id:96149eae-c99a-43e6-b616-97e9b690eb9b|ownerAccountId:b110042e-4a83-40bb-b54c-78376352ae0d": {
        "amount": "283.3400",
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "PEN",
        "id": "96149eae-c99a-43e6-b616-97e9b690eb9b",
        "lockedTimestamp": "2020-01-01T00:00:00.000Z",
        "note": "nostrum fugiat exercitationem",
        "ownerAccountId": "b110042e-4a83-40bb-b54c-78376352ae0d",
        "receiptId": null,
        "timestamp": "2020-01-01T00:00:00.000Z",
        "userId": "3f21fdd8-a0d9-409e-9a2a-b153d16f6bd5",
      },
      "id:96149eae-c99a-43e6-b616-97e9b690eb9b|ownerAccountId:bf922bf3-3fd7-4be5-b309-08e7e00e2fa3": {
        "amount": "-283.3400",
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "PEN",
        "id": "96149eae-c99a-43e6-b616-97e9b690eb9b",
        "lockedTimestamp": "2020-01-01T00:00:00.000Z",
        "note": "nostrum fugiat exercitationem",
        "ownerAccountId": "bf922bf3-3fd7-4be5-b309-08e7e00e2fa3",
        "receiptId": null,
        "timestamp": "2020-01-01T00:00:00.000Z",
        "userId": "8a5ebfc8-0bf8-49bd-9317-2ce5b13b8831",
      },
      "id:a0c1a92c-41a2-4d41-a06e-c368eee7f240|ownerAccountId:b110042e-4a83-40bb-b54c-78376352ae0d": {
        "amount": "510.6000",
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "OMR",
        "id": "a0c1a92c-41a2-4d41-a06e-c368eee7f240",
        "lockedTimestamp": "2020-01-01T00:00:00.000Z",
        "note": "quisquam mollitia at",
        "ownerAccountId": "b110042e-4a83-40bb-b54c-78376352ae0d",
        "receiptId": null,
        "timestamp": "2020-01-01T00:00:00.000Z",
        "userId": "3f21fdd8-a0d9-409e-9a2a-b153d16f6bd5",
      },
      "id:a0c1a92c-41a2-4d41-a06e-c368eee7f240|ownerAccountId:bf922bf3-3fd7-4be5-b309-08e7e00e2fa3": {
        "amount": "-510.6000",
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "OMR",
        "id": "a0c1a92c-41a2-4d41-a06e-c368eee7f240",
        "lockedTimestamp": "2020-01-01T00:00:00.000Z",
        "note": "quisquam mollitia at",
        "ownerAccountId": "bf922bf3-3fd7-4be5-b309-08e7e00e2fa3",
        "receiptId": null,
        "timestamp": "2020-01-01T00:00:00.000Z",
        "userId": "8a5ebfc8-0bf8-49bd-9317-2ce5b13b8831",
      },
      "id:ada8f5d5-b4f0-427d-8b01-49e30fb99e08|ownerAccountId:b110042e-4a83-40bb-b54c-78376352ae0d": {
        "amount": "-707.0800",
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "THB",
        "id": "ada8f5d5-b4f0-427d-8b01-49e30fb99e08",
        "lockedTimestamp": "2020-01-01T00:00:00.000Z",
        "note": "alias fugit iusto",
        "ownerAccountId": "b110042e-4a83-40bb-b54c-78376352ae0d",
        "receiptId": null,
        "timestamp": "2020-01-01T00:00:00.000Z",
        "userId": "3f21fdd8-a0d9-409e-9a2a-b153d16f6bd5",
      },
      "id:ada8f5d5-b4f0-427d-8b01-49e30fb99e08|ownerAccountId:bf922bf3-3fd7-4be5-b309-08e7e00e2fa3": {
        "amount": "707.0800",
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "THB",
        "id": "ada8f5d5-b4f0-427d-8b01-49e30fb99e08",
        "lockedTimestamp": "2020-01-01T00:00:00.000Z",
        "note": "alias fugit iusto",
        "ownerAccountId": "bf922bf3-3fd7-4be5-b309-08e7e00e2fa3",
        "receiptId": null,
        "timestamp": "2020-01-01T00:00:00.000Z",
        "userId": "8a5ebfc8-0bf8-49bd-9317-2ce5b13b8831",
      },
    },
  },
}
`;

exports[`debts.addBatch > functionality > debts are added - some are auto-accepted by the counterparty 1`] = `
{
  "added": {
    "debts": {
      "id:3fb62100-8408-48de-bb3a-f1e3f899ef83|ownerAccountId:b931b7cd-0bbb-4668-be2c-f59cfdecfddb": {
        "amount": "-794.6400",
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "UAH",
        "id": "3fb62100-8408-48de-bb3a-f1e3f899ef83",
        "lockedTimestamp": "2020-01-01T00:00:00.000Z",
        "note": "quisquam asperiores temporibus",
        "ownerAccountId": "b931b7cd-0bbb-4668-be2c-f59cfdecfddb",
        "receiptId": null,
        "timestamp": "2020-01-01T00:00:00.000Z",
        "userId": "a61eabf7-3766-45cc-884e-962e538d734c",
      },
      "id:6b8c3687-6276-48ef-b194-295365be669a|ownerAccountId:4adfd369-72a9-4631-903a-3353c3e5dba5": {
        "amount": "-835.0700",
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "LKR",
        "id": "6b8c3687-6276-48ef-b194-295365be669a",
        "lockedTimestamp": "2020-01-01T00:00:00.000Z",
        "note": "illo quidem nihil",
        "ownerAccountId": "4adfd369-72a9-4631-903a-3353c3e5dba5",
        "receiptId": "2474d40a-1505-46a5-a055-4495ed5f3b7b",
        "timestamp": "2020-01-01T00:00:00.000Z",
        "userId": "4670ef39-264b-4f7e-bfeb-13962a4e383c",
      },
      "id:6b8c3687-6276-48ef-b194-295365be669a|ownerAccountId:b931b7cd-0bbb-4668-be2c-f59cfdecfddb": {
        "amount": "835.0700",
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "LKR",
        "id": "6b8c3687-6276-48ef-b194-295365be669a",
        "lockedTimestamp": "2020-01-01T00:00:00.000Z",
        "note": "illo quidem nihil",
        "ownerAccountId": "b931b7cd-0bbb-4668-be2c-f59cfdecfddb",
        "receiptId": "2474d40a-1505-46a5-a055-4495ed5f3b7b",
        "timestamp": "2020-01-01T00:00:00.000Z",
        "userId": "e8537d1e-2f77-4a92-b31e-4188c399be5a",
      },
      "id:caffcfe2-fdb5-4e5a-8f19-f45b76e51e15|ownerAccountId:4adfd369-72a9-4631-903a-3353c3e5dba5": {
        "amount": "388.2900",
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "ILS",
        "id": "caffcfe2-fdb5-4e5a-8f19-f45b76e51e15",
        "lockedTimestamp": "2020-01-01T00:00:00.000Z",
        "note": "rerum tempore soluta",
        "ownerAccountId": "4adfd369-72a9-4631-903a-3353c3e5dba5",
        "receiptId": null,
        "timestamp": "2020-01-01T00:00:00.000Z",
        "userId": "4670ef39-264b-4f7e-bfeb-13962a4e383c",
      },
      "id:caffcfe2-fdb5-4e5a-8f19-f45b76e51e15|ownerAccountId:b931b7cd-0bbb-4668-be2c-f59cfdecfddb": {
        "amount": "-388.2900",
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "ILS",
        "id": "caffcfe2-fdb5-4e5a-8f19-f45b76e51e15",
        "lockedTimestamp": "2020-01-01T00:00:00.000Z",
        "note": "rerum tempore soluta",
        "ownerAccountId": "b931b7cd-0bbb-4668-be2c-f59cfdecfddb",
        "receiptId": null,
        "timestamp": "2020-01-01T00:00:00.000Z",
        "userId": "e8537d1e-2f77-4a92-b31e-4188c399be5a",
      },
      "id:f703d7a1-1162-440f-96e3-cddb7688bc47|ownerAccountId:1c1913a2-358a-42cf-96c1-97c9a12923e1": {
        "amount": "320.7700",
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "VEF",
        "id": "f703d7a1-1162-440f-96e3-cddb7688bc47",
        "lockedTimestamp": "2020-01-01T00:00:00.000Z",
        "note": "officiis qui quo",
        "ownerAccountId": "1c1913a2-358a-42cf-96c1-97c9a12923e1",
        "receiptId": null,
        "timestamp": "2020-01-01T00:00:00.000Z",
        "userId": "207afcd4-1bbf-4956-ac87-539df7c576e1",
      },
      "id:f703d7a1-1162-440f-96e3-cddb7688bc47|ownerAccountId:b931b7cd-0bbb-4668-be2c-f59cfdecfddb": {
        "amount": "-320.7700",
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "VEF",
        "id": "f703d7a1-1162-440f-96e3-cddb7688bc47",
        "lockedTimestamp": "2020-01-01T00:00:00.000Z",
        "note": "officiis qui quo",
        "ownerAccountId": "b931b7cd-0bbb-4668-be2c-f59cfdecfddb",
        "receiptId": null,
        "timestamp": "2020-01-01T00:00:00.000Z",
        "userId": "465229c7-e6a6-486b-a6bb-7ce4792a7b4b",
      },
    },
  },
}
`;

exports[`debts.addBatch > functionality > debts are added - user-provided timestamp 1`] = `
{
  "added": {
    "debts": {
      "id:90d46bf0-67b1-4316-8673-7e3dc542b214|ownerAccountId:26f5cb66-86fe-4a60-a7f2-466dbfbb1eab": {
        "amount": "155.4300",
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "XAF",
        "id": "90d46bf0-67b1-4316-8673-7e3dc542b214",
        "lockedTimestamp": "2020-01-01T00:00:00.000Z",
        "note": "aut sint libero",
        "ownerAccountId": "26f5cb66-86fe-4a60-a7f2-466dbfbb1eab",
        "receiptId": null,
        "timestamp": "2021-01-01T00:00:00.000Z",
        "userId": "b6206d08-839d-49f4-8930-608654e73331",
      },
      "id:abf6352d-3dfa-4796-87e4-2596648a6962|ownerAccountId:26f5cb66-86fe-4a60-a7f2-466dbfbb1eab": {
        "amount": "-37.9300",
        "created": "2020-01-01T00:00:00.000Z",
        "currencyCode": "HKD",
        "id": "abf6352d-3dfa-4796-87e4-2596648a6962",
        "lockedTimestamp": "2020-01-01T00:00:00.000Z",
        "note": "inventore occaecati voluptas",
        "ownerAccountId": "26f5cb66-86fe-4a60-a7f2-466dbfbb1eab",
        "receiptId": null,
        "timestamp": "2021-01-01T00:00:00.000Z",
        "userId": "b6f5e4b3-3e29-449f-b7c8-67a82910dedc",
      },
    },
  },
}
`;
